mmdet3d/core/bbox/coders/monoflex_bbox_coder.py

```diff
@@ -0,0 +1,387 @@
+        self.multibin = multibin
+        self.alpha_centers = alpha_centers
+        self.num_dir_bin = num_dir_bin
+
+        # output related
+        self.bbox_code_size = code_size
+        self.eps = 1e-3
+
+    def encode(self, locations, dimensions, orientations, input_metas):
+        """Encode CameraInstance3DBoxes by locations, dimensions, orientations.
+
+        Args:
+            locations (Tensor): Center location for 3D boxes.
+                (N, 3)
+            dimensions (Tensor): Dimensions for 3D boxes.
+                shape (N, 3)
+            orientations (Tensor): Orientations for 3D boxes.
+                shape (N, 1)
+            input_metas (list[dict]): Meta information of each image, e.g.,
+                image size, scaling factor, etc.
+
+        Return:
+            :obj:`CameraInstance3DBoxes`: 3D bboxes of batch images,
+                shape (N, bbox_code_size).
+        """
+
+        bboxes = torch.cat((locations, dimensions, orientations), dim=1)
+        assert bboxes.shape[1] == self.bbox_code_size, 'bboxes shape dose not'\
+            'match the bbox_code_size.'
+        batch_bboxes = input_metas[0]['box_type_3d'](
+            bboxes, box_dim=self.bbox_code_size)
+
+        return batch_bboxes
+
+    def decode(self, reg, points, labels, down_ratio, cam2imgs):
+        """Decode regression into 3D predictions.
+
+        Args:
+            reg (Tensor): Batch regression for each predict center2d point.
+                shape: (batch * K (max_objs), C)
+            points(Tensor): Batch projected bbox centers on image plane.
+                shape: (batch * K (max_objs) , 2)
+            labels (Tensor): Batch predict class label for each predict
+                center2d point.
+                shape: (batch * K (max_objs), )
+            down_ratio (int): The stride of feature map.
+            cam2imgs (Tensor): Batch images' camera intrinsic matrix.
+                shape: kitti (batch, 4, 4)  nuscenes (batch, 3, 3)
+
+        Return:
+            tuple(Tensor): The tuple has components below:
+                - locations (Tensor): Centers of 3D boxes.
+                    shape: (batch * K (max_objs), 3)
+                - dimensions (Tensor): Dimensions of 3D boxes.
+                    shape: (batch * K (max_objs), 3)
+                - orientations (Tensor): Orientations of 3D
+                    boxes.
+                    shape: (batch * K (max_objs), 1)
+        """
+
+        pred_offsets2d = reg[:, 4:6]
+        pred_dimensions_offsets3d = reg[:, 29:32]
+        pred_orientations = torch.cat((reg[:, 32:40], reg[:, 40:48]), dim=1)
+        # decode the pred residual dimensions to real dimensions
+        # (B* max_objs, 3)
+        pred_dimensions = self.decode_dimension(labels,
+                                                pred_dimensions_offsets3d)
+        # decode the depth with direct regression
+        pred_depth_offsets = reg[:, 48:49].squeeze(-1)
```

Comment:
Need some comments to specify which dimensions correspond to which predictions at the beginning

Reply:
I think we can refactor the code by adding another class: key2channels,  by using dimension str (like 'offsets2d')  to index the regression.